tools/slang-capability-generator/capability-generator-main.cpp

```diff
@@ -65,6 +77,86 @@ struct CapabilityDef : public RefObject
             return nullptr;
         return expr.conjunctions[0].atoms[0];
     }
+
+    const HashSet<CapabilityDef*>& getAbstractAtomsPresent() const
+    {
+        return abstractAtomsPresent;
+    }
+    void fillAbstractAtomsPresentFromCannonicalRepresentation()
+    {
+        bool alreadySetTarget = false;
+        HashSet<CapabilityDef*> firstTargetSet;
+
+        bool alreadySetStage = false;
+        HashSet<CapabilityDef*> firstStageSet;
+
+        HashSet<CapabilityDef*> abstractFound;
+
+        for (auto& c : canonicalRepresentation)
+        {
+
+            alreadySetTarget = false;
+            firstTargetSet.clear();
+            alreadySetStage = false;
+            firstStageSet.clear();
+
+            abstractFound.clear();
+
+            for (auto& atom : c)
+            {
+                for (auto& otherAbstractAtomsPresent : atom->abstractAtomsPresent)
+                {
+                    auto base = otherAbstractAtomsPresent->getAbstractBase();
+                    if (CapabilityDef::ptrOfTarget && base == CapabilityDef::ptrOfTarget)
+                    {
+                        if (!alreadySetTarget)
+                            firstTargetSet.add(otherAbstractAtomsPresent);
+                    }
+                    else if (CapabilityDef::ptrOfStage && base == CapabilityDef::ptrOfStage)
+                    {
+                        if (!alreadySetStage)
+                            firstStageSet.add(otherAbstractAtomsPresent);
+                    }
+
+                    abstractFound.add(otherAbstractAtomsPresent);
+                }
+                if (alreadySetTarget)
+                {
+                    for (auto i : firstTargetSet)
+                    {
+                        if (abstractFound.contains(i))
+                            continue;
+                        firstTargetSet.remove(i);
+                    }
+                }
+                if (alreadySetStage)
+                {
+                    for (auto i : firstStageSet)
+                    {
+                        if (!abstractFound.contains(i))
```

Comment:
Is this correct to have `!`?
The logic between firstStageSet and firstTargetSet appear to be almost identical exception this line.

Reply:
It seems like having `!` is the correct one and the other one should have `!` too.